daemons/ipa-kdb/ipa_kdb_kdcpolicy.c

```diff
@@ -44,13 +57,112 @@ ipa_kdcpolicy_check_as(krb5_context context, krb5_kdcpolicy_moddata moddata,
 
     krb5_klog_syslog(LOG_INFO, "IPA kdcpolicy: checking AS-REQ.");
 
-    return 0;
+    ipactx = ipadb_get_context(context);
+    if (!ipactx) {
+        kerr = KRB5_KDB_DBNOTINITED;
+        krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: Database not initiallized.");
+        goto done;
+    }
+
+    kerr = krb5_unparse_name(context, request->client, &principal);
+    if (kerr != 0) {
+        krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: krb5_unparse_name failed.");
+        goto done;
+    }
+
+    /* Filter out anonymous PKINIT principal since it doesn't have a LDAP entry */
+    if (strncmp(principal, "WELLKNOWN/ANONYMOUS@WELLKNOWN:ANONYMOUS", 40) == 0) {
+        goto done;
+    }
+
+    /* Get LDAP entry for the principal */
+    kerr = ipadb_fetch_principals(ipactx, KRB5_KDB_FLAG_ALIAS_OK, principal, &result);
+    if (kerr != 0) {
+        krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: ipadb_fetch_principals failed.");
+        goto done;
+    }
+
+    kerr = ipadb_find_principal(context, KRB5_KDB_FLAG_ALIAS_OK, result, &principal, &lentry);
+    if (kerr != 0) {
+        krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: ipadb_find_principal failed.");
+        goto done;
+    }
+
+    /* Get User Auth configuration. */
+    ua = ipadb_get_user_auth(ipactx, lentry);
+
+    /* Make sure string *auth_indicators has a length of 
+     * at least 1 so that strstr() won't broke */
+    if (*auth_indicators == NULL || strlen(*auth_indicators) == 0) {
+        if (!(ua & IPADB_USER_AUTH_PASSWORD)) {
+            kerr = KRB5KDC_ERR_POLICY;
+            *status = "Reinforced pre-authentication not allowed for this user.";
+        }
+        goto done;
+    }
+
+    /* For each auth indicator, see if it is allowed for that user */
+    if (strstr(*auth_indicators, "otp") != NULL) {
+        valid_indicators_set++;
+        if (!(ua & IPADB_USER_AUTH_OTP)) {
+            kerr = KRB5KDC_ERR_POLICY;
+            *status = "OTP pre-authentication not allowed for this user.";
+            goto done;
+        }
+    }
+
+    if (strstr(*auth_indicators, "radius") != NULL) {
+        valid_indicators_set++;
+        if (!(ua & IPADB_USER_AUTH_RADIUS)) {
+            kerr = KRB5KDC_ERR_POLICY;
+            *status = "RADIUS pre-authentication not allowed for this user.";
+            goto done;
+        }
+    }
+
+    if (strstr(*auth_indicators, "pkinit") != NULL) {
+        valid_indicators_set++;
+        if (!(ua & IPADB_USER_AUTH_PKINIT)) {
+            kerr = KRB5KDC_ERR_POLICY;
+            *status = "PKINIT pre-authentication not allowed for this user.";
+            goto done;
+        }
+    }
+
+    if (strstr(*auth_indicators, "reinforced") != NULL) {
+        valid_indicators_set++;
+        /* allow SPAKE/FAST even if only password is allowed */
+        if (!(ua & (IPADB_USER_AUTH_PASSWORD | IPADB_USER_AUTH_REINFORCED))) {
+            kerr = KRB5KDC_ERR_POLICY;
+            *status = "Password authentication not allowed for this user.";
```

Comment:
Ah I guess you mixed this error status with the one on top?
Also why do you check for IPADB_USER_AUTH_PASSWORD as well?
We may want to be  forcing the user to use only SPAKE but not pre-auth timestamp, and this check prevents that from working afaics.